persistent-postgresql/test/JSONTest.hs

```diff
@@ -40,431 +42,654 @@ insert' :: (MonadIO m, PersistStoreWrite backend, BaseBackend backend ~ SqlBacke
+          numFloatK <- insert' $ Number 0.0
+          numSmallK <- insert' $ Number 0.0000000000000000123
+          numFloat2K <- insert' $ Number 1.5
+          -- numBigFloatK will turn into 9876543210.123457 because JSON
+          numBigFloatK <- insert' $ toJSON (9876543210.123456789 :: Double)
+          let numbers = [ num0K, num1K, numBigK, numFloatK
+                        , numSmallK, numFloat2K, numBigFloatK ]
+
+          strNullK <- insert' $ String ""
+          strObjK <- insert' $ String "{}"
+          strArrK <- insert' $ String "[]"
+          strAK <- insert' $ String "a"
+          strTestK <- insert' $ toJSON ("testing" :: Text)
+          str2K <- insert' $ String "2"
+          strFloatK <- insert' $ String "0.45876"
+          let strings = [ strNullK, strObjK, strArrK, strAK
+                        , strTestK, str2K, strFloatK ]
+
+          arrNullK <- insert' $ Array $ V.fromList []
+          arrListK <- insert' $ toJSON [emptyArr,emptyArr,toJSON [emptyArr,emptyArr]]
+          arrList2K <- insert' $ toJSON [emptyArr,toJSON [Number 3,Bool False]
+                                        ,toJSON [emptyArr,toJSON [Object mempty]]
+                                        ]
+          arrFilledK <- insert' $ toJSON [Null, Number 4, String "b"
+                                         ,Object mempty, emptyArr
+                                         ,object [ "test" .= [Null], "test2" .= String "yes"]]
+          let lists = [ arrNullK, arrListK, arrList2K, arrFilledK ]
+
+          objNullK <- insert' $ Object mempty
+          objTestK <- insert' $ object ["test" .= Null, "test1" .= String "no"]
+          objDeepK <- insert' $ object ["c" .= Number 24.986, "foo" .= object ["deep1" .= Bool True]]
+          let objects = [ objNullK, objTestK, objDeepK ]
+          return GlobalKeys{..}
+
+
+data EdgeCaseKeys record =
+  EdgeCaseKeys { arrList3K  :: Key record
+               , arrList4K  :: Key record
+               , objEmptyK  :: Key record
+               , objFullK   :: Key record }
+
+data GlobalKeys record =
+  GlobalKeys { nulls   :: [Key record]
+             , bools   :: [Key record]
+             , numbers :: [Key record]
+             , strings :: [Key record]
+             , lists   :: [Key record]
+             , objects :: [Key record] }
+
 specs :: Spec
-specs = describe "postgresql's JSON operators behave" $ do
+specs =
+    describe "Testing JSON operators"
+    $ beforeAll setup
+    $ afterAll_ teardown $ do
+
+      -- Setup DB and get ahold of keys
+      GlobalKeys {..} <- runIO $ runConn_ $ globalKeys
+      let [ nullK ] = nulls
+      let [ boolTK, boolFK ] = bools
+      let [ num0K, num1K, numBigK, numFloatK, numSmallK, numFloat2K, numBigFloatK ] = numbers
+      let [ strNullK, strObjK, strArrK, strAK, strTestK, str2K, strFloatK ] = strings
+      let [ arrNullK, arrListK, arrList2K, arrFilledK ] = lists
+      let [ objNullK, objTestK, objDeepK ] = objects
```

Comment:
I'm *really* not a fan of getting the patterns like this, it's extremely fragile. It's more repetitive but it makes sense IMO to capture each key as a record field instead of constructing lists and then deconstructing them.

Reply:
Yeah, I agree with you. At first a field for each seemed like a lot of work, but I think you are right!